### DISCUSSION TITLE
Wrapping into custom components

### DISCUSSION THREAD

**Discussion by Ky6uk** [primefaces/community#860] [PrimeVue]
Hello.

Is it possible in a simple way to wrap existing components into custom components with some predefined things? For example, keep styles inside the component to make it reusable.

For example:

`CustomDataTable.vue`:

```vue
<template>
  <DataTable />
</template>

<style>
  .p-datatable-thead {
    color: magenta;
  }
</style>
```

`CustomColumn.vue`:

```vue
<template>
  <Column />
</template>
```

`App.vue`

```vue
<template>
  <CustomDataTable :value="tableData">
    <CustomColumn field="id" header="Id">
    <CustomColumn field="name" header="Name">
  <CustomDataTable>
</template>
```

**Comment by sergeydus** [comment]
same issue here, even tried to extend a component, but ts complains about classes

**Comment by Ky6uk** [comment]
In my particular case, I've tried to wrap `DataTable` and `Column` into `CustomDataTable` and `CustomColumn` components to keep custom stylings in the SFC, but it doesn't work at all (nothing rendered).

**Reply by sergeydus** [reply]
It sounds exactly like what I'm trying, except I'm trying to edit some styling inside the **Calendar** component instead.

I'd like it to behave IDENTICALLY to the default Calendar, with all the same props and template support except for the specific style change.

**Comment by sergeydus** [comment]
did you add 
```
import Column from 'primevue/column';
app.component('Column', Column);
```

in your menu.ts file?

**Comment by matthew-dean** [comment]
There really is no easy way to do this. There are a few problems:

1. In a TypeScript environment, [PrimeVue does not expose slots, props, and emits which can be used with `defineSlots`, `defineProps`, and `defineEmits` respectively](https://github.com/primefaces/primevue/issues/5405).
2. Vue itself makes extending components near impossible with `<script setup>`, because you can not easily "pass through" props etc. The Vue API _has_ this capability (mostly) via [the `extends` component property](https://vuejs.org/api/options-composition.html#extends), but not only does Vue not really support it for the composition API (they warn that setup won't be merged), and certainly doesn't support it _at all_ for `<script setup>`.

So, I have successfully extended, very easily, a library like Vuetify, but Vuetify's component structure is very straightforward, consistent, and easy to extend. PrimeVue's component structure really isn't extendable (as the short version), and even if it was, Vue's API direction is working strongly here against being able to abstract components. (Vue devs have filed numerous issues about this, but Vue seems to be maintained from the perspective of an author building all of their own components, even though a strong 3rd-party ecosystem exists.)

I personally hate React, but React has a long history with the concept of "higher-order components" and separation of concerns (which they kinda completely walked back when they shifted entirely to functional components and hooks, buuuut) which makes React components very very easy and trivial to extend, but Vue is a whole different beast with a lot of magic in the compiler (similar to Svelte) that makes extending components orders of magnitude more complicated, and you have to be extremely careful when designing Vue components to _make_ them extendable, and if this wasn't part of the initial component structure of PrimeVue (it wasn't), then it's very very hard to retro-fit because it usually requires a different API.

**Reply by derfl007** [reply]
> because you can not easily "pass through" props etc

You can with [Fallthrough Attributes](https://vuejs.org/guide/components/attrs.html), but that doesn't seem to work for slots and it also doesn't provide proper type hints
